### PR TITLE
Add support for instantiating the startup class

### LINKED ISSUE #19809: Startup class activation allow additional params
### Startup classes cannot be activated with additional constructor params

When using a startup class, suppose I want my constructor to have some arbitrary additional parameters- like an ILogger, or an IOptions object. At the moment this is not allowed because the host builder doesnt know how to satisfy those additional constructor parameters when it activates the startup class. However in other situations such as UseMiddleware you can supply those additional parameter values to be used when activating the class, at the point of registering it. You cant do that currently when registering a Startup class. This means if in startup.cs you want to use something like an ILogger that you created in program.cs, you have to resort to accessing it in a static field like Program.Logger etc.

### Describe the solution you'd like

The equivalent of 
`.UseStartup<Startup>(logger, foo, bar)`

When the startup class is activated, any constructor parameters that the host doesn't natively know how to resolve, will be matched from the additional object instances supplied at the point of registering the startup class. This is a pattern found elsewhere in certain places in the framework.

### PR DESCRIPTION
**Pull request by davidfowl**
- Adds an overload of UseStartup that takes a factory so users can control the instance creation. The factory is given the WebHostBuilderContext to expose access to configuration and IWebHostEnvironment.
- Make sure only one startup delegate runs, the last one registered.

Fixes #19809

### CODE REVIEW

File: src/Hosting/Hosting/src/GenericHost/GenericWebHostBuilder.cs [hunk 1]
```diff
@@ -210,13 +213,30 @@ public IWebHostBuilder UseStartup(Type startupType)
...
 
-        private void UseStartup(Type startupType, HostBuilderContext context, IServiceCollection services)
+        public IWebHostBuilder UseStartup(Func<WebHostBuilderContext, object> startupFactory)
+        {
+            // Clear the startup type
+            _startupObject = startupFactory;
+
+            _builder.ConfigureServices((context, services) =>
+            {
+                if (object.ReferenceEquals(_startupObject, startupFactory))
```
Comment (Tratcher) [inline src/Hosting/Hosting/src/GenericHost/GenericWebHostBuilder.cs:223]:
```suggestion
                // UseStartup can be called multiple times. Only run the last one.
                if (object.ReferenceEquals(_startupObject, startupFactory))
```

File: src/Hosting/Hosting/src/WebHostBuilderExtensions.cs [hunk 2]
```diff
@@ -61,6 +61,33 @@ private static IWebHostBuilder Configure(this IWebHostBuilder hostBuilder, Actio
...
         }
 
+        /// <summary>
+        /// Specify a factory that creates the startup instance to be used by the web host.
+        /// </summary>
+        /// <param name="hostBuilder">The <see cref="IWebHostBuilder"/> to configure.</param>
+        /// <param name="startupFactory">A delegate that specifies a factory for the startup class.</param>
+        /// <returns>The <see cref="IWebHostBuilder"/>.</returns>
+        public static IWebHostBuilder UseStartup(this IWebHostBuilder hostBuilder, Func<WebHostBuilderContext, object> startupFactory)
+        {
```
Comment (Tratcher) [inline src/Hosting/Hosting/src/WebHostBuilderExtensions.cs:71]: What about the app name?
https://github.com/dotnet/aspnetcore/blob/088380b45cad29f48635867526b954d8c86b84f7/src/Hosting/Hosting/src/WebHostBuilderExtensions.cs#L73-L75

**Review by Tratcher — COMMENTED** [review]
Looks good. Why is it still draft?

**Comment by davidfowl** [comment]
> Looks good. Why is it still draft?

I need to add a test and do API review.

File: src/Hosting/Hosting/src/WebHostBuilderExtensions.cs (hunk 2 above)
Comment (davidfowl) [reply src/Hosting/Hosting/src/WebHostBuilderExtensions.cs:71]: Good catch.

**Comment by halter73** [comment]
So this adds new public UseStartup method overloads that take a factory to HostingStartupWebHostBuilder and WebHostBuilderExtensions? Anything else? Do we have a good way to create an API diff yet?

**Comment by davidfowl** [comment]
> So this adds new public UseStartup method overloads that take a factory to HostingStartupWebHostBuilder and WebHostBuilderExtensions? Anything else? Do we have a good way to create an API diff yet?

That's the only new API.

File: src/Hosting/Hosting/src/WebHostBuilderExtensions.cs [hunk 3]
```diff
@@ -61,6 +61,37 @@ private static IWebHostBuilder Configure(this IWebHostBuilder hostBuilder, Actio
...
+            return hostBuilder
+                .ConfigureServices((context, services) =>
+                {
+                    services.AddSingleton(typeof(IStartup), sp =>
+                    {
+                        var instance = startupFactory(context) ?? new InvalidOperationException("The specified factory returned null startup instance");
+
+                        var hostingEnvironment = sp.GetRequiredService<IHostEnvironment>();
+
+                        return new ConventionBasedStartup(StartupLoader.LoadMethods(sp, instance.GetType(), hostingEnvironment.EnvironmentName, instance));
```
Comment (halter73) [inline src/Hosting/Hosting/src/WebHostBuilderExtensions.cs:91]: We should try as-casting the instance to IStartup first.

File: src/Hosting/Hosting/src/WebHostBuilderExtensions.cs [hunk 4]
```diff
@@ -61,6 +61,37 @@ private static IWebHostBuilder Configure(this IWebHostBuilder hostBuilder, Actio
...
 
+        /// <summary>
+        /// Specify a factory that creates the startup instance to be used by the web host.
+        /// </summary>
+        /// <param name="hostBuilder">The <see cref="IWebHostBuilder"/> to configure.</param>
+        /// <param name="startupFactory">A delegate that specifies a factory for the startup class.</param>
+        /// <returns>The <see cref="IWebHostBuilder"/>.</returns>
+        public static IWebHostBuilder UseStartup(this IWebHostBuilder hostBuilder, Func<WebHostBuilderContext, object> startupFactory)
+        {
+            var startupAssemblyName = startupFactory.GetMethodInfo().DeclaringType.GetTypeInfo().Assembly.GetName().Name;
```
Comment (pranavkm) [inline src/Hosting/Hosting/src/WebHostBuilderExtensions.cs:72]: null check `startupFactory`?

File: src/Hosting/Hosting/src/WebHostBuilderExtensions.cs [hunk 5]
```diff
@@ -61,6 +61,37 @@ private static IWebHostBuilder Configure(this IWebHostBuilder hostBuilder, Actio
...
+            {
+                return supportsStartup.UseStartup(startupFactory);
+            }
+
+            return hostBuilder
+                .ConfigureServices((context, services) =>
+                {
+                    services.AddSingleton(typeof(IStartup), sp =>
+                    {
+                        var instance = startupFactory(context) ?? new InvalidOperationException("The specified factory returned null startup instance");
```
Comment (pranavkm) [inline src/Hosting/Hosting/src/WebHostBuilderExtensions.cs:87]:
```suggestion
                        var instance = startupFactory(context) ?? new InvalidOperationException("The specified factory returned null startup instance.");
```

**Review by pranavkm — APPROVED** [review]
API looks fine by me. Feel free to send an email for broader visibility.

**Pull request merged**